File: removeDuplicates.js
const fs = require('fs');
const readline = require('readline').createInterface({
	input: process.stdin,
	output: process.stdout
})

var outputJSON = {
	participants: [],
	messages: [],
};

const removeDuplicateMessages = (filepath) => {
	// Read the file and parse the buffer to JSON
	const buffer = fs.readFileSync(filepath);
	const data = JSON.parse(buffer);
	
	console.log(`Messages in ${filepath}: ${data.messages.length}`);
	
	// Add the participants from the JSON if they do not exist
	data.participants.forEach(person => {
		if ( !outputJSON.participants.includes(person) ) {
			outputJSON.participants.push(person);
		}
	})
	
	// Add all messages to a set
	const uniqueMessages = new Set();
	data.messages.forEach(obj => {
		// Stringify before adding to set
		uniqueMessages.add(JSON.stringify(obj));
	})
	console.log(`Unique Messages in ${filepath}: ${uniqueMessages.size}\n`);
	
	// Push the unique messages to the output object
	uniqueMessages.forEach(message => {
		// Parse to JSON before pushing to output
		outputJSON.messages.push(JSON.parse(message));
	})
}

const writeToFile = (outputName) => {
	fs.writeFile(outputName, JSON.stringify(outputJSON, null, 2), 'utf8', () => {
		console.log(`Done writing file to: ${outputName}`)
	});
}

const promptUser = () => {

	readline.question(`Input the path to file or press 'n' to finish: `, (filepath) => {

		if ( filepath === 'n' ) {
			if ( outputJSON.messages.length ) {
				readline.question(`Specify the output file name: `, (outputName) => {
					writeToFile(outputName)
					readline.close();
				})
			}
			else {
				console.log(`The output array is empty!`)
				readline.close();
			}
			
		} else {
			removeDuplicateMessages(filepath);
			promptUser();
		}
	})

}

const main = () => {
	console.log(`Input most recent files first (message_1.json, then message_2.json, ...)\n`)
	promptUser();
}

main()

File: removeDuplicates.py
import json

outputJSON = {
	'participants': [],
	'messages': [],
}

def removeDuplicateMessages(filepath):

	with open(filepath, encoding="utf8") as json_file:
		data = json.load(json_file)

		print("Message in " + filepath + ": " + str(len(data['messages'])) )

		for person in data['participants']:
			if person['name'] not in outputJSON['participants']:
				outputJSON['participants'].append({'name': person['name']})
		
		uniqueMessages = set()
		for message in data['messages']:
			# Convert the JSON object to a string before adding to set
			uniqueMessages.add( json.dumps(message) )
		
		print("Unique Messages in " + filepath + ": " + str(len(uniqueMessages)) + "\n")
		
		for uniqueMessage in uniqueMessages:
			# Covert the JSON string back to JSON before appending to output
			outputJSON['messages'].append( json.loads(uniqueMessage) )
		
def writeToFile(outputFile):
	if not outputJSON['messages']:
		print("There are no messages in the output list!")
		return

	# Sort the messages by timestamp (most recent first)
	outputJSON['messages'].sort(key=lambda message: -message['timestamp_ms'])
	
	with open(outputFile, 'w', encoding='utf8') as output:
		json.dump(outputJSON, output, ensure_ascii=False, indent=4)
	
	print("Written to file: " + outputFile)

def main():

	while True:
		filepath = input("Input the path to file or press 'n' to finish: ")
		if filepath == 'n': break
		else: removeDuplicateMessages(filepath)
	
	outputFile = input("Please specify the output file name: ")
	writeToFile(outputFile)

if __name__ == "__main__":
	main()
